siwe/siwe.py

```diff
@@ -266,26 +283,37 @@ def verify(
 
         try:
             address = w3.eth.account.recover_message(message, signature=signature)
+        except ValueError:
+            address = None
         except eth_utils.exceptions.ValidationError:
             raise InvalidSignature
 
         if address != self.address:
-            raise InvalidSignature
-        #     if not check_contract_wallet_signature(message=self, provider=provider):
-        #         # TODO: Add error context
+            if not check_contract_wallet_signature(
+                address=self.address, message=message, signature=signature, w3=w3
+            ):
+                raise InvalidSignature
 
 
-def check_contract_wallet_signature(message: SiweMessage, *, provider: HTTPProvider):
+def check_contract_wallet_signature(
+    address: str, message: SignableMessage, signature: str, w3: Web3
+) -> bool:
     """
-    Calls the EIP-1271 method for Smart Contract wallets,
+    Calls the EIP-1271 method for Smart Contract wallets.
 
-    :param message: The EIP-4361 parsed message
-    :param provider: A Web3 provider able to perform a contract check.
+    :param address: The address of the contract
+    :param message: The EIP-4361 formatted message
+    :param signature: The EIP-1271 signature
+    :param w3: A Web3 provider able to perform a contract check.
     :return: True if the signature is valid per EIP-1271.
     """
-    raise NotImplementedError(
-        "siwe does not yet support EIP-1271 method signature verification."
-    )
+    contract = w3.eth.contract(address=address, abi=CONTRACT_ABI)
+    hash_ = _hash_eip191_message(message)
+    try:
+        response = contract.caller.isValidSignature(hash_, bytes.fromhex(signature[2:]))
+        return response.hex() == "1626ba7e"
```

Comment:
nit: add comment explaining what this magicvalue is (from eip: `// bytes4(keccak256("isValidSignature(bytes32,bytes)")
`)